### DISCUSSION TITLE
View from different schemas or db

### DISCUSSION THREAD

**Discussion by dom6688** [sqlalchemy/sqlalchemy#11184] [Usage Questions] [answered]
I have several schemas/dbs with different bind_key, like:

```
SQLALCHEMY_BINDS = {'st1': "oracle+cx_oracle://../st1",
                    'st2': "oracle+cx_oracle://../st2"
                    }
```
I need to create a view using tables with different bind_key.

```
class Table1(Base): table = Table( 'st1', Base.metadata, Column('id', Integer, primary_key=True), autoload_with=db.get_engine(bind='st1'), info={'bind_key': 'st1'} )

class Table2(Base): table = Table( 'st2', Base.metadata, Column('id', Integer, primary_key=True), autoload_with=db.get_engine(bind='st2'), info={'bind_key': 'st2'} )

class TestView(Model): selectable= select(Table1.name, Table2.name).select_from(Table1).join(Table2, Table2.ik==Table1.ik)) bind_key="st1" table = create_view( name='test_view', selectable=selectable, metadata=db.Model.metadata ), )

```
There is a problem mostly with metadata object which can't find table. Error: "ORA-00942: table or view does not exist .."

Is it possible to create a view without using "raw" SQL query?

**Comment by CaselIT** [comment, marked as answer]
Hi,

If the dame db can access both tables, and is seems it does since you are trying to use a single query with both tables, then consider using a single engine and use schemas with that engine

**Reply by dom6688** [reply]
Unfortunately, I have different databases and engines. The query mentioned is the one I would like to use, and I wrote it only as an example and to show that it doesn't work.
I can create a view using "raw" SQL query (without sqlAlchemy), but I need to use SqlAlchemy for it.

I tried to use also one metadata object with reflection and automap. But then I try to create a view, I need to select the engine and here I will have problems with duplicates of all reflected tables...

**Reply by CaselIT** [reply]
Using a single engine with schemas is likely the easiest solution here.

I think you can map an orm model to a raw sql query using `text()`

**Reply by dom6688** [reply]
Sorry for the many stupid questions.
It may be easier to show a simple test example. This is one of many attempts. And it doesn't work.

Here I'm trying to reflect all tables in one metadata object so that select works. Without this, tables from one of the databases can't be found
```
app = Flask(__name__)

app.config["SQLALCHEMY_DATABASE_URI"] = "postgresql://test1:123123@localhost:5432/test1"
app.config["SQLALCHEMY_BINDS"] = {
    "test2": "postgresql://test2:123123@localhost:5432/test2",
    "test1": "postgresql://test1:123123@localhost:5432/test1",
}
db = SQLAlchemy(app)
Session1 = sessionmaker(bind=db.get_engine(bind="test1"))
Session2 = sessionmaker(bind=db.get_engine(bind="test2"))

class User(db.Model):
    __bind_key__ = "test1"
    __tablename__ = "user"
    id = Column(Integer, primary_key=True)
    name = Column(String(80), unique=True, nullable=False)

class UserSecond(db.Model):
    __bind_key__ = "test2"
    __tablename__ = "user_second"
    id = Column(Integer, primary_key=True)
    surname = Column(String(80), unique=True, nullable=False)

def add_user(index, name, surname):
    new_user = User(id=index, name=name)
    with Session1() as session:
        session.add(new_user)
        session.commit()

    with Session2() as session:
        new_user_surname = UserSecond(id=index, surname=surname)
        session.add(new_user_surname)
        session.commit()

if __name__ == "__main__":
    db.drop_all()
    db.create_all()

    metadata = MetaData(bind=None)
    metadata.reflect(db.get_engine(bind="test1"))
    metadata.reflect(db.get_engine(bind="test2"))

    Base = automap_base(metadata=metadata)
    Base.prepare()

    add_user(1, "Al", "Mil")
    add_user(2, "Winny", "Col")

    selectable = (
        select(Base.classes.user.name, Base.classes.user_second.surname)
        .select_from(Base.classes.user)
        .join(
            Base.classes.user_second,
            Base.classes.user.id == Base.classes.user_second.id,
        )
    )
    create_view("test_view", selectable, Base.metadata)
    metadata.create_all(db.get_engine())
    app.run(debug=True)


```

**Reply by CaselIT** [reply]
It's really hard to follow what you are trying to do, since now you are using postgresql, while before you were using oracle. The db in use may change things 

For postgresql, those seem two different databases, and you can't query them with a single statement as far as I know: https://wiki.postgresql.org/wiki/FAQ#How_do_I_perform_queries_using_multiple_databases.3F

**Reply by dom6688** [reply]
Yes, you're absolutely right. I apologize for the confusion. It was just a dummy example to explain a task

Is this possible for Oracle with sqlAlchemy? I know that with Raw queries it can be done.
This task is originally for an Oracle database and I need to do it there.

**Reply by CaselIT** [reply]
Can you post an example of sql query that works with oracle?
If that's something like
```sql
select db1.foo.col, db2.bar.col
from db1.foo
join db2.bar on db1.foo.id = db2.bar.id
```

Those are schemas for sqlalchemy and you can use a single engine indicating that table `foo` is in schema `db1` while table `bar` is in schema `db2`.

If you need to absolutely use two engines for whatever reason, I think you will need to either manually indicate that one of tables has a schema and then use the engine connected to the table that doesn't have one.
I don't know how flask sqlalchemy organizes things here, but if it uses two metadata you can do `tableCopy = table.to_metadata(other_engine_metadata, schema='<the-schema-you-want>')`

Overall using a single engine is the easiest solution here
